users/vieting/experiments/switchboard/ctc/feat/network_helpers/specaug_configurable.py

```diff
@@ -0,0 +1,300 @@
+    _, indices = tf.math.top_k(z, num if isinstance(num, int) else tf.reduce_max(num))
+    # indices should be sorted, and of shape (batch,num), entries (int32) in [0,dim)
+    if isinstance(num, int):
+        for i in range(num):
+            x = _mask(
+                x,
+                batch_axis=batch_axis,
+                axis=axis,
+                pos=indices[:, i],
+                max_amount=max_dims,
+                sorted_indices=sorted_indices,
+            )
+    else:
+        _, x = tf.while_loop(
+            cond=lambda i, _: tf.less(i, tf.reduce_max(num)),
+            body=lambda i, x: (
+                i + 1,
+                tf.where(
+                    tf.expand_dims(tf.expand_dims(tf.less(i, num), axis=-1), axis=-1),
+                    _mask(
+                        x,
+                        batch_axis=batch_axis,
+                        axis=axis,
+                        pos=indices[:, i],
+                        max_amount=max_dims,
+                        sorted_indices=sorted_indices,
+                    ),
+                    x,
+                ),
+            ),
+            loop_vars=(0, x),
+        )
+    return x
+
+
+def transform(data, network, **config):
+    import tensorflow as tf
+
+    x = data.placeholder
+    step = network.global_train_step
+    current_epoch = tf.cast(step / config["steps_per_epoch"], tf.int32)
+
+    specaug_params = config["specaug_params"]
+
+    # Determine if we should use sorting
+    use_sorting = tf.logical_and(
+        config["enable_sorting"], tf.greater_equal(current_epoch, config["sorting_start_epoch"])
+    )
+
+    # Get filter indices (sorted or unsorted)
+    filter_layer = network.layers["features"].subnetwork.layers["conv_h_filter"].output.placeholder
+    num_filters = network.layers["features"].subnetwork.layers["conv_l"].output.shape[-1]
+
+    def get_sorted_indices():
+        sorted_filter_indices = sort_filters_by_center_freq(filter_layer)
+        sorted_indices = tf.stack(
+            [sorted_filter_indices * num_filters + filter_idx for filter_idx in range(num_filters)]
+        )
+        return tf.reshape(tf.transpose(sorted_indices), (-1,))
+
+    def get_unsorted_indices():
+        return tf.range(tf.shape(x)[data.feature_dim_axis])
+
+    sorted_indices = tf.cond(use_sorting, get_sorted_indices, get_unsorted_indices)
+
+    def get_masked():
+        x_masked = x
+        time_mask_max_size = tf.gather(specaug_params["time_mask_max_size"], current_epoch)
+        time_mask_max_num = tf.gather(specaug_params["time_mask_max_num"], current_epoch)
+        freq_mask_max_num = tf.gather(specaug_params["freq_mask_max_num"], current_epoch)
+        freq_mask_max_size = tf.gather(specaug_params["freq_mask_max_size"], current_epoch)
+        time_mask_max_proportion = tf.cast(
+            tf.math.floor(config["time_mask_max_proportion"] * tf.cast(tf.shape(x)[data.time_dim_axis], tf.float32)),
```

Comment:
Why is this cast needed?

Reply:
I got type errors without it. I think because the output of math.floor has the same type as the input and since `config["freq_mask_max_proportion"]` is a float the type would be wrong